src/cls/cmpxattr/server.cc

```diff
@@ -0,0 +1,169 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2024 IBM Corp.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ */
+
+#include "objclass/objclass.h"
+#include "ops.h"
+#include "common/errno.h"
+CLS_VER(1,0)
+CLS_NAME(cmpxattr)
+
+using namespace cls::cmpxattr;
+
+// returns negative error codes or 0/1 for failed/successful comparisons
+template <typename T>
+static int compare_values(Op op, const T& lhs, const T& rhs)
+{
+  switch (op) {
+  case Op::EQ:  return (lhs == rhs);
+  case Op::NE:  return (lhs != rhs);
+  case Op::GT:  return (lhs > rhs);
+  case Op::GTE: return (lhs >= rhs);
+  case Op::LT:  return (lhs < rhs);
+  case Op::LTE: return (lhs <= rhs);
+  default:      return -EINVAL;
+  }
+}
+
+static int compare_values_u64(Op op, uint64_t lhs, const bufferlist& value)
+{
+  // empty values compare as 0 for backward compat
+  uint64_t rhs = 0;
+  if (value.length()) {
+    try {
+      // decode existing value as rhs
+      auto p = value.cbegin();
+      using ceph::decode;
+      decode(rhs, p);
+    } catch (const buffer::error&) {
+      // failures to decode existing values are reported as EIO
+      return -EIO;
+    }
+  }
+  return compare_values(op, lhs, rhs);
+}
+
+static int compare_value(Mode mode, Op op, const bufferlist& input,
+			 const bufferlist& value)
+{
+  switch (mode) {
+  case Mode::String:
+    return compare_values(op, input, value);
+  case Mode::U64:
+    try {
+      // decode input value as lhs
+      uint64_t lhs;
+      auto p = input.cbegin();
+      using ceph::decode;
+      decode(lhs, p);
+      return compare_values_u64(op, lhs, value);
+    } catch (const buffer::error&) {
+      // failures to decode input values are reported as EINVAL
+      return -EINVAL;
+    }
+  default:
+    return -EINVAL;
+  }
+}
+
+static int cmp_vals_set_vals(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
```

Comment:
as discussed, the function should return and error code or zero to mark sucess.
in the out buffer we can return either 0/1 to indicate if the comparisson was evaluated to false/true (in case of sucess), or, maybe how many values were set (any non zero value indicate comparisson was "true")

Reply:
done